tests/unit/test_charm.py

```diff
@@ -479,15 +480,122 @@ def test_mongodb_departed_non_leader_does_nothing(self, reconfigure):
+        # preset values
+        self.harness.charm._need_replica_set_reconfiguration = True
+        # cannot hard set this since its used in other locations
+        self.harness.charm.unit.name = "mongodb/1"
+        departing_unit.return_value = "mongodb/1"
+        self.harness.charm._primary = "mongodb/1"
+
+        # add peer unit
+        rel = self.harness.charm.model.get_relation("mongodb")
+        key_values = {"private-address": "127.4.5.6"}
+        self.harness.add_relation_unit(rel.id, "mongodb/1")
+        self.harness.update_relation_data(rel.id, "mongodb/1", key_values)
+
+        # remove unit to trigger event
+        self.harness.remove_relation_unit(rel.id, "mongodb/1")
+
+        # verify we call the step down method
+        primary_step_down.assert_called()
+
+    @patch_network_get(private_address="1.1.1.1")
+    @patch("charm.MongodbOperatorCharm._primary")
+    @patch("os.environ.get")
+    @patch("mongod_helpers.MongoDB.primary_step_down")
+    @patch("charm.MongodbOperatorCharm._single_mongo_replica")
+    @patch("charm.MongodbOperatorCharm._need_replica_set_reconfiguration")
+    @patch("mongod_helpers.MongoDB.reconfigure_replica_set")
+    def test_on_mongodb_relation_departed_primary_removed_failure(
+        self,
+        mongodb_reconfigure,
+        need_reconfiguration,
+        single_replica,
+        primary_step_down,
+        departing_unit,
+        primary,
+    ):
+        """Test removing primary replica leads to failure.
+
+        Verifies relation_departed operations for when the departing unit is a primary and
+        a failure occurs in attempying to step down the primary replica.
+        """
+        # preset values
+        self.harness.charm._need_replica_set_reconfiguration = True
+        self.harness.charm.unit.name = "mongodb/0"
+        departing_unit.return_value = "mongodb/0"
+        self.harness.charm._primary = "mongodb/0"
+        self.harness.set_leader(True)
+
+        exceptions = [
+            ConnectionFailure("connection error message"),
+            ConfigurationError("configuration error message"),
+            OperationFailure("operation error message"),
+        ]
+
+        # check that each exception is handled properly
+        for exception in exceptions:
+            primary_step_down.side_effect = exception
+
+            # add peer unit
+            rel = self.harness.charm.model.get_relation("mongodb")
+            rel_id = rel.id
+            key_values = {"private-address": "127.4.5.6"}
+            self.harness.add_relation_unit(rel_id, "mongodb/1")
+            self.harness.update_relation_data(rel_id, "mongodb/1", key_values)
+
+            # remove unit to trigger event
+            self.harness.remove_relation_unit(rel_id, "mongodb/1")
+
+            # verify waiting status
+            self.assertTrue(isinstance(self.harness.charm.unit.status, WaitingStatus))
+
+            # verify we don't reconfigure
+            mongodb_reconfigure.assert_called()
```

Comment:
yes it should be... so its concerning that this test passes.... 😬

Reply:
so actually after doing some digging the reason why this passes is because when we do:
```
            # add peer unit
            rel = self.harness.charm.model.get_relation("mongodb")
            rel_id = rel.id
            key_values = {"private-address": "127.4.5.6"}
            self.harness.add_relation_unit(rel_id, "mongodb/1")
            self.harness.update_relation_data(rel_id, "mongodb/1", key_values)
```

 reconfigure gets called since we have to first add this to the replica set. 


Rather than adding a check here for re-configure, I think we can remove this check.

My reasoning is that we verify that we don't update the hosts with :
```
            # verify we don't update
            self.assertEqual(
                self.harness.charm._replica_set_hosts,
                ["127.4.5.6", "1.1.1.1"],
            )
```

Reply:
Thanks for the explanation, Mia! It's clear to me.